### PR TITLE
implemented order model

### PR DESCRIPTION
**Pull request by foulwinter09**

### CODE REVIEW

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        order.setTotal(total);
+        orderRepository.save(order);
+        orderItemRepository.saveAll(order.getOrderItems());
+
+        shoppingCart.getCartItems().clear();
+        shoppingCartRepository.save(shoppingCart);
+
+        return orderMapper.toDto(order);
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]: let's divide this method into private helpful

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/model/OrderItem.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/model/OrderItem.java]:
```suggestion
```

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+    private static Order makeOrder(String shippingAddress, User user) {
+        Order order = new Order();
+        order.setUser(user);
+        order.setStatus(Order.Status.PENDING);
+        order.setOrderDate(LocalDateTime.now());
+        order.setShippingAddress(shippingAddress);
+        order.setDeleted(false);
+        return order;
+    }
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]: why do you need static?

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        return shoppingCart;
+    }
+
+    private User getUser(Long userId) {
+        return userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException("User not found"));
+    }
+
+    private BigDecimal calculateTotalAndAddOrderItems(ShoppingCart shoppingCart, Order order) {
+        BigDecimal total = BigDecimal.ZERO;
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]: no Single Responsibility method

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java (hunk 3 above)
Comment (foulwinter09) [reply src/main/java/mate/academy/service/impl/OrderServiceImpl.java]: i used "extract method" from intellij and this func makes methods static by default, i just do not noticed that.

File: src/main/java/mate/academy/controller/OrderController.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    @PatchMapping("/{orderId}")
+    @Operation(summary = "Update order status",
+            description = "Allows an admin to update the status of an order.")
+    @ApiResponses({
+            @ApiResponse(responseCode = "200", description = "Order status updated successfully"),
+            @ApiResponse(responseCode = "400", description = "Invalid status"),
+            @ApiResponse(responseCode = "401", description = "Unauthorized"),
+            @ApiResponse(responseCode = "404", description = "Order not found")
+    })
+    public void updateOrderStatus(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/controller/OrderController.java]: Update endpoint should return dto

File: src/main/java/mate/academy/dto/order/OrderDto.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.Set;
+
+public record OrderDto(
+        @NotNull Long id,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/order/OrderDto.java]: Do we need validate response?

File: src/main/java/mate/academy/dto/order/OrderItemDto.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package mate.academy.dto.order;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+import java.math.BigDecimal;
+
+public record OrderItemDto(@NotNull Long id,
+                           @NotNull Long orderId,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/order/OrderItemDto.java]: same

File: src/main/java/mate/academy/dto/order/PlaceOrderRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.dto.order;
+
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+
+@Data
+public class PlaceOrderRequestDto {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/order/PlaceOrderRequestDto.java]: Why is not record?

File: src/main/java/mate/academy/dto/order/UpdateOrderStatusRequestDto.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.dto.order;
+
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+import mate.academy.model.Order;
+
+@Data
+public class UpdateOrderStatusRequestDto {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/order/UpdateOrderStatusRequestDto.java]: same

File: src/main/java/mate/academy/model/Order.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false)
+    private Status status;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/model/Order.java]: you can initialise status here

File: src/main/java/mate/academy/model/Order.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/model/Order.java]: you can initialise date now here

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+        order.setTotal(total);
+        saveOrderAndOrderItems(order);
+
+        clearShoppingCart(shoppingCart);
+
+        return orderMapper.toDto(order);
+    }
+
+    @Transactional(readOnly = true)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]: remove, we need @Transactional only for place order method

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        return orderItemRepository.findByOrderId(orderId, pageable).stream()
+                .map(orderItemMapper::toDto)
+                .toList();
+    }
+
+    @Transactional(readOnly = true)
+    @Override
+    public OrderItemDto getOrderItem(Long orderId, Long itemId) {
+        OrderItem orderItem = orderItemRepository.findByOrderIdAndId(orderId, itemId)
+                .orElseThrow(() -> new EntityNotFoundException("Order item not found"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]: add id to message

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        return orderItemMapper.toDto(orderItem);
+    }
+
+    @Transactional
+    @Override
+    public void updateOrderStatus(Long orderId, Order.Status status) {
+        Order order = orderRepository.findById(orderId)
+                .orElseThrow(() -> new EntityNotFoundException("Order not found"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]: same

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+        orderRepository.save(order);
+    }
+
+    private Order makeOrder(String shippingAddress, User user) {
+        Order order = new Order();
+        order.setUser(user);
+        order.setStatus(Order.Status.PENDING);
+        order.setOrderDate(LocalDateTime.now());
+        order.setShippingAddress(shippingAddress);
+        order.setDeleted(false);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]:
```suggestion
        order.setShippingAddress(shippingAddress);
```

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+        return order;
+    }
+
+    private ShoppingCart getShoppingCart(Long userId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]:
```suggestion
        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId).orElseThrow(...
```

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+        if (shoppingCart == null || shoppingCart.getCartItems().isEmpty()) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]:
```suggestion
        if (shoppingCart.getCartItems().isEmpty()) {
```

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+            throw new EntityNotFoundException("Shopping cart is empty");
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]:
```suggestion
            throw new OrderProcessingException(....
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**